### PR TITLE
[ARM64/Windows] Add JIT_Stelem_Ref helper

### LINKED ISSUE #13590: [Arm64/Windows] Add JIT_Stelem_Ref helper
Implement the Windows equivalent of #13588 

category:cq
theme:helpers
skill-level:expert
cost:small

### PR DESCRIPTION
**Pull request by pgodeq**
Adding Windows equivalent of JIT_Stelem_Ref #13588

Fixes #13590

### CODE REVIEW

**Comment by pgodeq** [comment]
@sdmaclea @jkotas @janvorli @jashook PTAL
@dotnet/arm64-contrib FYI

**Comment by sdmaclea** [comment]
@dotnet/arm64-contrib PTAL

File: src/vm/arm64/asmhelpers.asm [hunk 1]
```diff
@@ -1386,5 +1391,99 @@ CallHelper2
...
+   LEAF_END 
+
+; ------------------------------------------------------------------
+; __declspec(naked) void F_CALL_CONV JIT_Stelem_Ref_NotExactMatch(PtrArray* array,
+;                                                       unsigned idx, Object* val)
+;   x12 = array->GetArrayElementTypeHandle()
+;
+    NESTED_ENTRY JIT_Stelem_Ref_NotExactMatch
+    PROLOG_SAVE_REG_PAIR           fp, lr, #-0x80!
+    SAVE_ARGUMENT_REGISTERS        sp, 0x10
```
Comment (sdmaclea) [inline src/vm/arm64/asmhelpers.asm]: This will save x0-x8, but we really only need x0-x2.

Comment (pgodeq) [reply src/vm/arm64/asmhelpers.asm]: There seems to be a restriction on saving registers x0-x2 with macro PROLOG_SAVE_REG_PAIR. 
It accepts register numbers x19 to x31 only.

Comment (sdmaclea) [reply src/vm/arm64/asmhelpers.asm]: @janvorli @jkotas Is this a known restriction?

Comment (jkotas) [reply src/vm/arm64/asmhelpers.asm]: `PROLOG_SAVE_REG_PAIR` generates unwind info. Unwind info needs to be generate for non-volatile registers only (ie x19 - x31).

If you need to save scratch registers, you can same them after prolog. Or if you really need to save them during prolog, you can use `PROLOG_NOP`.

File: src/vm/arm64/asmhelpers.asm [hunk 2]
```diff
@@ -1386,5 +1391,101 @@ CallHelper2
...
+    add     x1, sp, #16             ; x1 = &array
+    add     x0, sp, #32             ; x0 = &val
+
+    bl      ArrayStoreCheck ; ArrayStoreCheck(&val, &array)
+
+DoWrite        
+	ldp     x0, x1, [sp], #16
+	ldr     x2, [sp], #32	
+    EPILOG_RESTORE_REG_PAIR           fp, lr, #0x48!
+    b JIT_Stelem_DoWrite
```
Comment (jkotas) [inline src/vm/arm64/asmhelpers.asm]: Should this be `EPILOG_BRANCH` to generate the right unwind info?

Comment (pgodeq) [reply src/vm/arm64/asmhelpers.asm]: I agree. Thanks for pointing it. I am correcting it.

**Comment by pgodeq** [comment]
@dotnet-bot test Tizen armel Cross Debug Build

**Pull request merged**